### ISSUE TITLE
Inconsistent behaviour related to OpCompositeConstant

### ISSUE THREAD

**Issue by stefanomil** [KhronosGroup/SPIRV-Tools#3576]
**TL;DR:** The constant manager does not accept some integer vector
constants that the validator considers valid and accepts other ones
that should not be considered valid.

----

Given the following instructions:

```
%1 = OpTypeInt 32 0
%2 = OpTypeInt 32 1
%3 = OpConstant %1 1
%4 = OpConstant %2 1
%5 = OpTypeVector %1 4
%6 = OpConstantComposite %5 %4 %4 %4 %4
%7 = OpConstantComposite %5 %3 %4 %3 %4
```

The constant manager does not record vector %7 as being a constant.
In fact, the comment above the CreateConstant function declaration
in constants.h says:

"
  When creating VectorType Constant instance, the components must be
  scalars of the same type, either Bool, Integer or Float.
"
which is consistent with the behaviour, because %3 and %4 have
different types (%3 is an unsigned integer, %4 is signed). Weirdly,
%6 is recognised, even if the type of the vector is unsigned int
and the type of the components is signed int.

However, the validator does not complain about %7, which can be
used as a vector of unsigned integers and works as expected.

Adding the following lines:

```
%8 = OpTypeFloat 32
%9 = OpConstant %8 0
%10 = OpConstantComposite %5 %9 %9 %9 %9
```

the constant manager accepts %10 as a vector constant while the
validator (rightly) complains that the component type and the
vector type are not compatible.

This is probably because, in constants.cpp, the first element type
rather than the component type of the vector is checked against to
verify that all types are the same.

From lines 269-275 of constants.cpp:

```c++
const auto* component_type = components.front()->type();
if (!std::all_of(components.begin(), components.end(),
                 [&component_type](const Constant* c) {
                   if (c->type() == component_type) return true;
                   return false;
                 }))
  return nullptr;
```

----
- %10 being recognised seems an error by the constant manager
- %7 not being accepted by the constant manager but accepted by the
  validator shows an inconsistency. Which one is right?
 
The specification for OpCompositeConstant says:

"
Result Type must be a composite type, whose top-level
members/elements/components/columns have the same
type as the types of the Constituents.
"

**Comment by afd** [comment]
@stefanomil Nice TL;DR, but I'd put that at the start of your message :)  (You can edit.)

@alan-baker / @s-perron - I would value your input here.

Regarding "%10 being recognised seems an error by the constant manager": my view on an invalid module is that we don't care too much what the constant manager does: I think it's reasonable to only care about the constant manager's behaviour on SPIR-V that the validator accepts.

Regarding "%7 not being accepted by the constant manager but accepted by the validator shows an inconsistency. Which one is right?" - the validator is corect that this is valid.  Whether the constant manager is wrong depends on the goals of the constant manager.  Is it supposed to manage all constants, or is it deliberately restricted to constants that are in some sense "nice"?  (i.e. that do not mix sign in the way they are constructed)

**Comment by s-perron** [comment]
>  I think it's reasonable to only care about the constant manager's behaviour on SPIR-V that the validator accepts.

I agree with that.

>  Is it supposed to manage all constants, or is it deliberately restricted to constants that are in some sense "nice"?

It should manage all constants.  However, I am surprised that %7 is legal.  The spec says "Result Type must be a composite type, whose top-level members/elements/components/columns have the same type as the types of the Constituents."  In this case, the member type does not match the constituent type.  However, I am not an expert on the spec.  I'll defer the Alan on that.

**Comment by afd** [comment]
Thanks @s-perron.  @alan-baker what do you think re the legality or not of %7?

**Comment by alan-baker** [comment]
I would have thought the validator would catch this. At first glance the code is there to cover this case. I do agree it should be invalid.

**Comment by afd** [comment]
@stefanomil Can you upload a spirv binary that exhitits the problem and confirm that spirv-val (when compiled from the latest source) indeed doesn't complain about it?

**Comment by stefanomil** [comment]
[Here](https://github.com/KhronosGroup/SPIRV-Tools/files/4987432/shader.txt) it is. It still does not complain. I changed the extension to .txt so that I could upload the file, but it actually is the binary. Running spirv-dis on it should gives:

```
; SPIR-V
; Version: 1.5
; Generator: Khronos SPIR-V Tools Assembler; 0
; Bound: 13
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main"
               OpExecutionMode %main OriginUpperLeft
               OpSource ESSL 310
               OpName %main "main"
       %void = OpTypeVoid
       %uint = OpTypeInt 32 0
        %int = OpTypeInt 32 1
     %uint_1 = OpConstant %uint 1
      %int_1 = OpConstant %int 1
     %v4uint = OpTypeVector %uint 4
          %9 = OpConstantComposite %v4uint %uint_1 %uint_1 %uint_1 %uint_1
         %10 = OpConstantComposite %v4uint %uint_1 %int_1 %uint_1 %int_1
         %11 = OpTypeFunction %void
       %main = OpFunction %void None %11
         %12 = OpLabel
               OpReturn
               OpFunctionEnd
```
